source/credentials_provider_ecs.c

```diff
@@ -0,0 +1,636 @@
+        AWS_LOGF_TRACE(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p): tls context not provided, initializing a new one",
+            (void *)provider);
+        struct aws_tls_ctx_options tls_options;
+        aws_tls_ctx_options_init_default_client(&tls_options, allocator);
+        impl->ctx = aws_tls_client_ctx_new(allocator, &tls_options);
+
+        if (!impl->ctx) {
+            AWS_LOGF_ERROR(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "(id=%p): failed to create a tls context with error %s",
+                (void *)provider,
+                aws_error_debug_str(aws_last_error()));
+            aws_tls_ctx_options_clean_up(&tls_options);
+            goto on_error;
+        }
+
+        impl->owns_ctx = true;
+        aws_tls_connection_options_init_from_ctx(&impl->connection_options, impl->ctx);
+        struct aws_byte_cursor host = options->host;
+        if (aws_tls_connection_options_set_server_name(&impl->connection_options, allocator, &host)) {
+            AWS_LOGF_ERROR(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "(id=%p): failed to create a tls connection options with error %s",
+                (void *)provider,
+                aws_error_debug_str(aws_last_error()));
+            goto on_error;
+        }
+    }
+
+    struct aws_socket_options socket_options;
+    AWS_ZERO_STRUCT(socket_options);
+    socket_options.type = AWS_SOCKET_STREAM;
+    socket_options.domain = AWS_SOCKET_IPV4;
+    socket_options.connect_timeout_ms = (uint32_t)aws_timestamp_convert(
+        ECS_CONNECT_TIMEOUT_DEFAULT_IN_SECONDS, AWS_TIMESTAMP_SECS, AWS_TIMESTAMP_MILLIS, NULL);
+
+    struct aws_http_connection_manager_options manager_options;
+    AWS_ZERO_STRUCT(manager_options);
+    manager_options.bootstrap = options->bootstrap;
+    manager_options.initial_window_size = ECS_RESPONSE_SIZE_LIMIT;
+    manager_options.socket_options = &socket_options;
+    manager_options.host = options->host;
+    manager_options.port = impl->owns_ctx ? 443 : 80;
+    manager_options.max_connections = 2;
+    manager_options.shutdown_complete_callback = s_on_connection_manager_shutdown;
+    manager_options.shutdown_complete_user_data = provider;
+    manager_options.tls_connection_options = impl->owns_ctx ? &(impl->connection_options) : NULL;
+
+    impl->function_table = options->function_table;
+    if (impl->function_table == NULL) {
+        impl->function_table = &s_default_function_table;
+    }
+
+    impl->connection_manager = impl->function_table->aws_http_connection_manager_new(allocator, &manager_options);
+    if (impl->connection_manager == NULL) {
+        goto on_error;
+    }
+    if (options->auth_token.len != 0) {
+        impl->auth_token = aws_string_new_from_array(allocator, options->auth_token.ptr, options->auth_token.len);
+    }
+    impl->path_and_query =
+        aws_string_new_from_array(allocator, options->path_and_query.ptr, options->path_and_query.len);
```

Comment:
need to check results of `aws_string_new_xxx()`
could fail if we ran out of memory